Feat: Dafny format file.dfy and IDE extension (#2399)
dafny-lang/dafny: docs/DafnyRef/UserGuide.md

working directory. A command-line argument not matching a known option is considered a filepath, and likely one
with an unsupported suffix, provoking an error message.
- Files containing dafny code must have a `.dfy` suffix.

> - There must be at least one `.dfy` file (except in the case of `dafny format`, see the [Dafny format section](#sec-formatting)) 

Reviewer: I strongly prefer that dafny format not be an exception. The argument could be a directory, or an option like --repo. But an empty argument list is often an accident (e.g. dafny format *.dy), and this one would start a lot of work that might modify files.

Author: Your example is not a problem for the current dafny formatter, it reports as expected

```
> ./Binaries/Dafny format *.dy
*** Error: 'C:\Users\mimayere\Documents\dafny\*.dy': Filename extension '.dy' is
 not supported. Input files must be Dafny programs (.dfy) or supported auxiliary
 files ()

I also tried with `unexisting*.dfy` and it throws an exception (might not be good but at least it's complaining it cannot read the file)

So here are my thoughts
* I think there are two schools for formatters.
* Some formatting tools I got inspiration from do the same:
  * [dotnet format](https://github.com/dotnet/format) formats all .cs files in the current directory
  * [scalafmt](https://scala-cli.virtuslab.org/docs/commands/fmt/) has the same behavior, it formats all Scala files
  * [cargo fmt](https://github.com/rust-lang/rustfmt) does the same for rust
* Some formatting tools do your suggested approach and don't support zero arguments
  * [go fmt <path>](https://pkg.go.dev/cmd/gofmt), without path it process the stdin
  * [tsfmt -r](https://www.npmjs.com/package/typescript-formatter) does recursive formatting only if there is a project file
  * [JIndent](http://www.newforms-tech.com/products/jindent/components/console) requires an explicit directory.

I remember you worked a lot in Java before, so that's how I understand your strong opinion. I come from a Scala background so I'm used to the first option and I never had an issue as you suggested. At worst, I could kill the formatter.
I personally prefer the first approach because it's a simple action to do - no big deal to format files anyway, if people used the formatter before, their files should be already formatted using it.
I'm ok adding support for directory names as well, down the road. For this first version, I prefer to stick to simplicity. Files => format one file. Nothing => format the directory recursively.

Reviewer: Ah, that behavior (in bash anyway) depends on whether `nullglob` is set. So some users will see `*.dy` stay `*.dy` (if there are no .dy files) and some will see it be replaced by an empty list.

So I'm still strongly for requiring a specific argument (though allowing it to be stdin if there is no argument is OK).
I think it much safer as the command is altering files (recursively!) in place. 

The problem with your 'simplicity' is that it would be a breaking change to then require a directory in the future. If you can currently do the current directory recursively, it seems a very simple change to require a directory and format an arbitrary directory instead of the current directory. And writing `dafny format .` is hardly harder and much clearer than `dafny format`

> - There must be at least one `.dfy` file (except in the case of `dafny format`, see the [Dafny format section](#sec-formatting)) 

Reviewer: ```suggestion
- There must be at least one `.dfy` file (except when using `--stdin` or in the case of `dafny format`, see the [Dafny format section](#sec-formatting)) 
```

For example, if a program has parsing errors, verification or compilation will
not be attempted.

### 25.6.1. The Dafny formatter

> Dafny supports a formatter, which for now only changes the indentation of every line in a Dafny file, so that it conforms
> to a specific model that matches the format of most of what the Dafny team wrote in the past.

Reviewer 2: That the formatting style matches historical Dafny code is an interesting fact, but IMHO we should be more prescriptive here. If we intend that the formatter enforces "idiomatic" Dafny style, we should say something along the lines of:

```suggestion
Dafny supports a formatter, which for now only changes the indentation of every line in a Dafny file, so that it conforms
to the idiomatic Dafny code formatting style.
```

(or perhaps a bit weaker to indicate that the formatter only captures the indentation of the idiomatic Dafny code formatting style)

Author: > (or perhaps a bit weaker to indicate that the formatter only captures the indentation of the idiomatic Dafny code formatting style)

I like the way you phrased it, so I applied your changes.

For the formatter to work, the file should be parsed correctly by Dafny.

There are three ways to use the formatter:

* `dafny format file1.dfy file2.dfy ...` formats each file provided in arguments.
* `dafny format /print:- file1.dfy file2.dfy ...` format each file but instead of writing files, output their formatted content in stdout

> * `dafny format` Look up for all the Dafny files recursively in the current repository and format them.

Author: I have the "-check" option ready in a future commit. I just need to add some tests.

> * `dafny format` Look up for all the Dafny files recursively in the current repository and format them.

Author: I just pushed a commit that includes the `--check` option as you suggested. It was actually a feature already asked by some other users.

#### 25.5.1.9. `dafny format` {#sec-dafny-format}

Dafny supports a formatter, which for now only changes the indentation of every line in a Dafny file, so that it conforms
to the idiomatic Dafny code formatting style.
For the formatter to work, the file should be parsed correctly by Dafny.

There are four ways to use the formatter:

* `dafny format file1.dfy file2.dfy ...` formats each file provided in arguments.
* `dafny format --print:- file1.dfy file2.dfy ...` format each file but instead of writing files, output their formatted content in stdout

> * `dafny format` Look up for all the Dafny files recursively in the current repository and format them.

Reviewer: No -- being in a repo is a special case. I suggest the arguments be files or directories, and use something 
like --repo if one really needs to format the repo (without having to write the directory for it)

Author: What do you mean by "needs to format the repo without having to write the directory for it"? One could always explicitly pass "." which is shorter than "--repo".

Author: That's a correct guess ! Thanks for clarifying your point.

* `dafny format --check` with the same options as above. It will print a message
  and a non-zero return code if at least one file is not the same as

>   its formatted version through `dafny format`

Author: 1. I changed the "--print:-" option to a boolean "--print" which was very easy thanks to the new command line
2. dotnet format already does that, and our formatter is pretty stable. I would like to get more opinions on that @robin-aws @keyboardDrummer and @alex-chew  ? Vote 🅰️ if you want to enforce "." and 🅱️  if you think `dafny format` is enough
3. --check does not modify any file, and tells which files are not correctly formatted, like this.
```
$ ./Binaries/Dafny.exe format --check Source/DafnyCore/AST/Formatting.dfy
The file Formatting.dfy needs to be formatted
Error: 1 file need formatting.

$ ./Binaries/Dafny.exe format --check Source/DafnyCore/AST/Formatting.dfy
All files correctly formatted
```
Could we please keep diffs for an upcoming PR based on needs?

>   its formatted version through `dafny format`

Reviewer 3: I think `dafny format` should not try to format everything if no argument is passed, since that seems too implicit of a behavior. Regarding formatting a directory, can't a user use shell wildcards to pass many individual files to the CLI? Why do we need first class directory support?

I do think it should accept individual files or a project file (when it becomes available)

>   its formatted version through `dafny format`

Reviewer: To format all dafny files recursively under a directory would take combining with a `find` call. I think it better to have some options to recursively format all dafny files under some directory.

>   its formatted version through `dafny format`

Author: You can always use "ls" and "xargs" to provide a specific list of files to format to Dafny.

I'm optimizing for the two most common cases:
* Formatting all the dafny files recursively in a repository as part of a commit process: `dafny format .` (I adopted your suggestion @davidcok  of accepting folders and `dafny format` now returns some usage help)
* Formatting specific dafny files.

>   its formatted version through `dafny format`

Author: To do this, there is already the following command:

```
dafny format --stdin --print
```
so I don't think it's worth adding a special case for "-". First and foremost, no one complained that stdin was not supported in the new Dafny version so I'm pretty sure it won't be a pain point for now, and if it is later, someone could always suggest a PR, but I would be skeptical to support ad-hoc use cases.

>   its formatted version through `dafny format`

Reviewer: Agree - since --print now takes no argument, we don't need -. But what is the default output  when using --stdin? I think it might as well be stdout (not anything like stdin.dfy). One can always redirect it where one wants.

>   its formatted version through `dafny format`

Author: Currently, the use of --stdin requires the use of either --check or --print, otherwise it returns an error and explanation.

>   its formatted version through `dafny format`

Reviewer: That's OK, at least until user feedback says differently.

#### 25.5.1.9. `dafny format` {#sec-dafny-format}

Dafny supports a formatter, which for now only changes the indentation of every line in a Dafny file, so that it conforms
to the idiomatic Dafny code formatting style.
For the formatter to work, the file should be parsed correctly by Dafny.

There are four ways to use the formatter:

* `dafny format file.dfy [file2.dfy ...]` formats the given Dafny files, altering the files in place.

> * `dafny format folder` formats all the Dafny files contain in the given folders, altering the files in place. For example, `dafny format .` formats all the Dafny files recursively in the current folder.

Reviewer: Can one have multiple folders? Mix them with files?

Author: Yes to both. Do you have a suggestion on how to word it efficiently?

* `dafny format --print file.dfy` formats each file but instead of altering the files, output the formatted content to stdout
* `dafny format --check file.dfy` does not alter files. It will print a message concerning which files need formatting.

You can also use `--stdin` instead of providing a file to format a full Dafny file from the standard input.

> In any case, each version of `dafny format` returns a non-zero return code

Reviewer: Which error code is being used to indicate unformatted input? Something different than other error codes I hope.
I would have suggested a non-zero error code only on
- command-line errors (1, as for dafny)
- parse/resolution errors (2, as for dafny)
- unformatted input, but only for --check -- perhaps 5, to be different than other kinds of errors

Author: I'm using `ExitValue.DAFNY_ERROR`, which is 2. Switched to a new value (probably 5), and only if --check is present.

There are four ways to use the formatter:

* `dafny format file1.dfy file2.dfy ...` formats each file provided in arguments, altering the files in place.
* `dafny format --print file1.dfy file2.dfy ...` format each file but instead of altering the files, output the formatted content to stdout

> * `dafny format` Look up for all the Dafny files recursively in the current repository and format them.

Reviewer: But I still want to see the community's take on this behavior -- I really want there to be some argument required for there to be such a big change.

> * `dafny format --check` with the same options as above. It will print a message

Reviewer: Still not clear whether this option changes any files. I vote for no change. Rather print out a list of which files would change, preferably with a diff of the changes.

Author: > There is no way to format material from stdin. I suggest `dafny format` take its input from stdin (and prints to stdout by default). And then have a directory required to format files in a directory.

I added support for the option "--stdin" which existed before for Dafny.

Author: I refactored this support into a separate PR as well
https://github.com/dafny-lang/dafny/pull/3429
